packages/flutter_tools/lib/src/base/deferred_component.dart

```diff
@@ -0,0 +1,291 @@
+  }
+
+  /// Parses the loading unit manifests from the [Environment.outputDir] of the latest
+  /// gen_snapshot/assemble run.
+  ///
+  /// This will read all existing loading units for every provided abi. If no abis are
+  /// provided, loading units for all abis will be parsed.
+  static List<LoadingUnit> parseGeneratedLoadingUnits(Environment env, {List<String> abis}) {
+    final List<LoadingUnit> loadingUnits = <LoadingUnit>[];
+    final List<FileSystemEntity> files = env.outputDir.listSync(recursive: true);
+    while (files.isNotEmpty) {
+      if (files.last is File) {
+        final File file = files.last as File;
+        // Determine if the abi is one we build.
+        bool matchingAbi = abis == null;
+        if (abis != null) {
+          for (final String abi in abis) {
+            if (file.parent.path.endsWith(abi)) {
+              matchingAbi = true;
+              break;
+            }
+          }
+        }
+        if (!file.path.endsWith('manifest.json') || !matchingAbi) {
+          files.removeLast();
+          continue;
+        }
+        loadingUnits.addAll(parseLoadingUnitManifest(file));
+      }
+      files.removeLast();
+    }
+    return loadingUnits;
+  }
+
+  /// Parses loading units from a single loading unit manifest json file.
+  ///
+  /// Returns an empty list if the manifestFile does not exist or is invalid.
+  static List<LoadingUnit> parseLoadingUnitManifest(File manifestFile) {
+    if (!manifestFile.existsSync()) {
+      return <LoadingUnit>[];
+    }
+    // Read gen_snapshot manifest
+    final String fileString = manifestFile.readAsStringSync();
+    Map<String, dynamic> manifest;
+    try {
+      manifest = jsonDecode(fileString) as Map<String, dynamic>;
+    } on FormatException catch (e) {
+      globals.printError(e.toString());
+    }
+    final List<LoadingUnit> loadingUnits = <LoadingUnit>[];
+    // Setup android source directory
+    if (manifest != null) {
+      for (final dynamic loadingUnitMetadata in manifest['loadingUnits']) {
+        final Map<String, dynamic> loadingUnitMap = loadingUnitMetadata as Map<String, dynamic>;
+        if (loadingUnitMap['id'] == 1) {
+          continue; // Skip base unit
+        }
+        loadingUnits.add(LoadingUnit(
+          id: loadingUnitMap['id'] as int,
+          path: loadingUnitMap['path'] as String,
+          libraries: List<String>.from(loadingUnitMap['libraries'] as List<dynamic>)),
+        );
+      }
+    }
+    return loadingUnits;
+  }
+}
+
+/// Utility method to copy and rename the required .so shared libs from the build output
+/// to the correct component intermediate directory.
+///
+/// The [DeferredComponent]s passed to this method must have had loading units assigned.
+/// Assigned components are components that have determined which loading units contains
+/// the dart libraries it has via the DeferredComponent.assignLoadingUnits method.
+Depfile copyDeferredComponentSoFiles(
```

Comment:
I would put this in the build_system library and keep base free of these deps if possible